### PR TITLE
QA-168 Implement Test cases collector and Matcher

### PR DESCRIPTION
**Pull request by madlabman**

### CODE REVIEW

**Comment by skhomuti** [comment]
Unit test needed to demonstrate how it works

File: src/pytest_allure_spec_coverage/plugin.py [hunk 1]
```diff
@@ -9,4 +9,166 @@
...
+    """Get scenario links from pytest.Item"""
+
+    return safe_get_marker(item, ScenariosMatcher.MARKER_NAME).args
+
+
+def allure_labels(scenario: Scenario) -> List[Label]:
+    """Make labels for Allure from the given scenario"""
+
+    labels = ("parentSuite", "suite", "subSuite")
+    values = [p.name for p in scenario.parents]
```
Comment (skhomuti) [inline src/pytest_allure_spec_coverage/plugin.py:53]: use `display_name` instead of name

File: src/pytest_allure_spec_coverage/plugin.py [hunk 2]
```diff
@@ -9,4 +9,166 @@
...
+            self._report_scenario(scenario)
+
+    def _report_scenario(self, scenario: Scenario) -> None:
+        fake_uuid = uuid4()
+        fake_result = TestResult(
+            uuid=fake_uuid,
+            name=scenario.name,
+            description=scenario.display_name,
+            status=Status.UNKNOWN,
+            labels=allure_labels(scenario),
```
Comment (skhomuti) [inline src/pytest_allure_spec_coverage/plugin.py:101]: The scenario also has `link` attr
Place `Link` allure object to `links` attr of TestResult

**Review by skhomuti — CHANGES_REQUESTED** [review]

File: src/pytest_allure_spec_coverage/hooks.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# limitations under the License.
+
+"""Plugin hooks"""
+
+from typing import Mapping
+
+from pytest_allure_spec_coverage.models.collector import Collector
+
+
+def pytest_register_spec_collectors(collectors: Mapping[str, Collector]) -> None:  # pylint: disable=unused-argument
```
Comment (dgusakov) [inline src/pytest_allure_spec_coverage/hooks.py:20]: Add hookspec decorator

File: src/pytest_allure_spec_coverage/plugin.py [hunk 4]
```diff
@@ -9,4 +9,166 @@
...
+def scenario_links(item: Item) -> Iterable[str]:
+    """Get scenario links from pytest.Item"""
+
+    return safe_get_marker(item, ScenariosMatcher.MARKER_NAME).args
+
+
+def allure_labels(scenario: Scenario) -> List[Label]:
+    """Make labels for Allure from the given scenario"""
+
+    labels = ("parentSuite", "suite", "subSuite")
```
Comment (dgusakov) [inline src/pytest_allure_spec_coverage/plugin.py:52]: Use custom labels for implemented scenarios and custom + default for non implemented ones
@skhomuti  add custom labels names

Comment (dgusakov) [reply src/pytest_allure_spec_coverage/plugin.py:52]: It seems to me that a list of these labels should be configurable from the `pyproject.toml` config

File: src/pytest_allure_spec_coverage/plugin.py [hunk 5]
```diff
@@ -9,4 +9,166 @@
...
+    reporter: AllureReporter
+
+    scenarios: Iterable[Scenario] = field(default_factory=tuple)
+    implemented: Iterable[str] = field(default_factory=tuple)
+
+    @property
+    def missed(self) -> Iterable[Scenario]:
+        """Not implemented scenarios"""
+
+        return (scenario for scenario in self.scenarios if scenario.link not in self.implemented)
```
Comment (dgusakov) [inline src/pytest_allure_spec_coverage/plugin.py:78]: It is not correct to use full link here.
@skhomuti pls add a corresponding attribute to the SpecCollector so we can use it here

**Review by dgusakov — COMMENTED** [review]
Can we also add simple tests for the implemented functionality?

File: src/pytest_allure_spec_coverage/plugin.py (hunk 1 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/plugin.py]: +

File: src/pytest_allure_spec_coverage/plugin.py (hunk 2 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/plugin.py:127]: Done.

File: src/pytest_allure_spec_coverage/hooks.py (hunk 3 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/hooks.py:23]: Done.

File: src/pytest_allure_spec_coverage/plugin.py (hunk 5 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/plugin.py]: Done.

File: src/pytest_allure_spec_coverage/plugin.py [hunk 6]
```diff
@@ -9,4 +9,194 @@
...
+    if len(values) > 3:
+        itv = slice(2, None)
+        values[itv] = [".".join(values[itv])]
+    return [Label(label, value) for label, value in zip(labels, values)]
+
+
+def allure_links(scenario: Scenario) -> List[Link]:
+    """Make links for Allure from the given scenario"""
+
+    return [Link(url=scenario.link, name="Scenario", type=LinkType.LINK)]
```
Comment (dgusakov) [inline src/pytest_allure_spec_coverage/plugin.py:67]:
```suggestion
    return [Link(url=scenario.link, name="Spec Link", type=LinkType.LINK)]
```

File: src/pytest_allure_spec_coverage/plugin.py (hunk 4 above)
Comment (dgusakov) [reply src/pytest_allure_spec_coverage/plugin.py:56]: Still relevant

File: src/pytest_allure_spec_coverage/plugin.py (hunk 6 above)
Comment (dgusakov) [reply src/pytest_allure_spec_coverage/plugin.py:67]: Moreover, it should be configurable too

Comment (skhomuti) [reply src/pytest_allure_spec_coverage/plugin.py:67]: Also, link is optionally

File: src/pytest_allure_spec_coverage/plugin.py [hunk 7]
```diff
@@ -9,4 +9,194 @@
...
+
+        for scenario in self.missed:
+            self._report_scenario(scenario)
+
+    def _report_scenario(self, scenario: Scenario) -> None:
+        fake_uuid = uuid4()
+        fake_result = TestResult(
+            uuid=fake_uuid,
+            name=scenario.name,
+            description=scenario.display_name,
```
Comment (skhomuti) [inline src/pytest_allure_spec_coverage/plugin.py:125]:
```suggestion
            name=scenario.display_name,
```
`display name` is not a description

File: src/pytest_allure_spec_coverage/plugin.py [hunk 8]
```diff
@@ -9,4 +9,194 @@
...
+            status=Status.UNKNOWN,
+            labels=allure_labels(scenario),
+            links=allure_links(scenario),
```
Comment (skhomuti) [inline src/pytest_allure_spec_coverage/plugin.py:128]:
```suggestion
            links=allure_links(scenario) if scenario.link else None,
```

File: src/pytest_allure_spec_coverage/plugin.py (hunk 7 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/plugin.py:125]: Removed.

File: src/pytest_allure_spec_coverage/plugin.py (hunk 4 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/plugin.py]: Added ConfigProvider.

File: src/pytest_allure_spec_coverage/plugin.py (hunk 6 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/plugin.py]: Done.

File: src/pytest_allure_spec_coverage/plugin.py (hunk 8 above)
Comment (madlabman) [reply src/pytest_allure_spec_coverage/plugin.py]: Done.

**Review by dgusakov — APPROVED** [review]

**Review by dgusakov — APPROVED** [review]

**Review by skhomuti — DISMISSED** [review]

**Review by skhomuti — APPROVED** [review]

**Review by dgusakov — APPROVED** [review]

**Pull request merged**